### DISCUSSION TITLE
Web client version with webAssembly

### DISCUSSION THREAD

**Discussion by PauloBarbeiro** [fontforge/fontforge#5298] [Ideas]
Hello,

Is there any plan/initiative for a web-browser distribution?

I am researching webAssembly and I wish to work in bringing a project to the browser environment, and FontForge is a grate project.
If there is any plan/initiative, could someone tell me how to get involved?
Otherwise, I wish to ask to some of the core maintainers, what are the core parts of the system. That would be helpful to start my learning. :) like, what composes the system's core, the ui components, and the IO interfaces.

All the best,

**Comment by iorsh** [comment]
Hi, currently FontForge is somewhat stale, but there are grassroot efforts to modernize its UI by porting it to GTK3. The current implementation uses custom UI toolkit (25 years old) with GDK under the hood, and it's really difficult to maintain or enhance.

The initial GTK3 effort is done by @Omnikron13 at https://github.com/Omnikron13/fontforge/tree/gtk3-hybrid and by me (@iorsh at https://github.com/iorsh/fontforge/tree/recent_files)

I know nothing about webAssembly, so can't judge if it can be employed with current UI or needs some adjustments. I believe any contribution is welcome, but I'm not a maintanier, so can't speak in their name.

**Reply by skef** [reply]
I recommend separating the question of GTK3 from the question of browser support, or at least putting the questions in that order. A GTK3 port could help with in-browser use but I doubt there would be support for just abandoning the desktop client, so the first stage would be just getting GTK3 working in the desktop context.

I haven't looked at the above links, but one question is how complete these ports are. FontForge has many, many interfaces, including a large number of modal dialogs. Were _all_ of those ported to GTK3, or just a subset of them that the developer thought were relevant? Does the bitmap font window work, along with its dialogs? If not, you would have to get buy-in on abandoning those facilities, and it's very unclear how one would get buy-in on that given the distributed and largely silent nature of the user community. 

You might be better off attempting to fork and to gain separate support for your fork. However, be warned: there have been many past partial forks of FontForge that have failed.

**Comment by skef** [comment]
I also noticed that a recent push to the first link above enables c++20 modules. I think a successful FontForge port will need to stay further back from the bleeding edge -- it is used in a lot of contexts and it's not safe to assume that all of them will have recent compilers. C++17 would be safer.

**Reply by Omnikron13** [reply]
re; things like using C++20 features, these efforts to try and modernise the UI (and ultimately the codebase in general) are very early, very experimental, and very rough at present. Actually integrating meaningful changes into a usable fork, let alone bring things into the mainline are likely quite some way off anyway, but indeed the more bleeding edge things are really simply probing around the direction that reorganising and reworking the codebase could go over the coming years (and indeed decades) than a serious consideration for the short term.

**Comment by myoresh** [comment]
See also #4812.
@lancejpollard, at the moment I don't know if there is anyone around here with sufficient knowledge of wasm.
CC @jimmywarting

**Comment by warting** [comment]
I just did it. POC available at https://onlinefontconverter.com

https://github.com/premex-ab/fontforge-wasm
